man/testhost.1
.\" $Revision: 3.10.0 $
.TH TESTHOST 1
.SH NAME
testhost - test the status of an NNTP news server
.SH SYNOPSIS
.I testhost
.BI hostname
[
.BI -a | -n
date time
.BI | -o
]
[
.BI -M
]
[
.BI -s | -S
filename
]
[
.BI -e | -E
filename
]
[
.BI -N
port_number
]
[
.BI -U
userid
]
[
.BI -P
password
]
[
.BI -Q
]
[
.BI -l
phrase_file
]
[
.BI -T
timeout
]
[
.BI -d
]
[
.BI -q
]
[
.BI -z
]

.SH OPTIONS
-a     Get the active list from hostname

\-d

This option tells testhost get the descriptions of the newsgroups on the
remote server by sending the 'list newsgroups' command.  The remote server
may or may not support this command.

-e | -E filename

These options will send all error messages (normally displayed on stderr), to 
an alternate file.  The lower case version, -e, will send the error messages
to the compiled-in default defined in suck_config.h.  The default is suck.errlog.
The upper case version, -E, requires the filename parameter.  All error messages
will then be sent to this file.

\-l phrase_file

This option tells testhost to load in an alternate phrase file, instead of using
the built-in messages.  This allows you to have testhost print phrases in another
language, or to allow you to customize the messages without re-building.
See the "FOREIGN LANGUAGE PHRASES" in suck.1 for more details.

-n date time     

Get the newgroups created on the host since the date and
time specified. The date must be in YYMMDD format and the time must be in
HHMMSS format.

-N port_number

This option will tell testhost to use an alternate NNRP port number when connecting
to the host, instead of the default, 119.

\-q 

This option tells testhost to not display the connection and announcement messages,
rather only display the results of the actual command run.

-s | -S filename

These options will send all status messages (normally displayed on stdout), to
an alternate file.  The lower case version, -s, will send the status messages
to the compiled-in default defined in suck_config.h.  The default is /dev/null,
so no status messages will be displayed.  The upper case version, -S, requires
the filename parameter.  All status messages will then be sent to this file.

\-T

This option overrides the compiled-in TIMEOUT value. This is how long testhost
waits for data from the remote host before timing out and aborting.

\-U userid

\-P password

These two options let you specify a userid and password, if your NNTP server
requires them.

\-Q

This option tells testhost to use the environment variable NNTP_USER & NNTP_PASS
to specify a userid and password, if your NNTP server requires them.  This option
is provided so that the userid & password can't be seen with the ps command, a
potential security problem.

\-z

This options tells testhost to use SSL to talk to the remote server, if testhost
was compiled with SSL.

.SH DESCRIPTION
.I Testhost
will query a NNTP news server, specified by
.I hostname.

The hostname may optionally include the port number in the form
.BI Host:Port.  If the port number is included, the port number specified
by the -N option will be ignored.

The default action is to issue the 
.I help
command to the server, to see what software it is running
and what commands it accepts.
.PP
If the
.I -a
option is used, testhost will display the server's
active history list.  If the 
.I -n date time
option is used, testhost will display all new groups created on the server
since the date and time specified.  If the
.I -o
option is used, then testhost will display the overview format, which is
what the XOVER command returns.
.PP
If the
.I -M
option is used, the command will be preceded with the "mode reader" command,
which might be needed by some servers.

.SH EXIT VALUES
0 on success, -1 on failure.
.de R$
This is revision \\$3, \\$4.
..
.SH "SEE ALSO"
suck(1), rpost(1), lpost(1).


CHANGELOG

CHANGES
=======
-- Changes from suck-4.3.1 -> Suck-4.3.2

Wherever a hostname is specified on the command line, the form Host:Port
may be used to specify a port number.  If this form is used, the -N port
option will be ignored.

configure.in - add -with-perl-exe option to specify an
	alternate path to the perl exe.

suck.c, killfile.c - get_a_chunk(), get_chunk_mem()
	changed to handle lines beginning with .
	correctly in stdout mode.

suck.c, rpost.c - main() - error handling if host isn't
	specified on command line.

* - various patches from Debian folks
configure.in - check for configdata.h in inn include directory

both.c - sgetline() - changes to fix SSL bug with the select,
	connect_to_nntphost() - add code to handle
	host:port syntax.   This fix also allows you to
	specify a port number for the local host.

suck.c - to handle SIGINT in addition to SIGTERM,
	build_command() fix in potential buffer overflow

testhost.c - added -Q option, to allow user to specify
	NNTP authentiation via env variables.

March 28, 2003 - Suck-4.3.2 released

-- Changes from Suck-4.3.0 -> Suck-4.3.1

* as of this version, my e-mail address has changed
to bobyetman@sucknews.org

* 	- added stuff to autoconf to handle the Db,
	history, Perl & SSL stuff in the Makefile.

* both.h - added tests for socks.h

* rpost.c - added -Q option, to allow user
	to specify NNTP authentication via
	environment variables
         - do_batch() - add test to not append
	prefix if article name is an Inn token.

* suck.c - added -Q option, to allow user
	to specify NNTP authentication via
	environment variables

	- build_command() - changed sprintfs
	to snprintf() to fix potential overflow
	problems.

	-- get_one_article() - fix bug where
	plist would be null when doing authentication
	and there was only one article to download.

        - main() - fix bug where if reconnect after
	dedupe, and it failed, you would try to send
	the quit command to an invalid file descriptor.

Sep 24, 2002, Suck-4.3.1 released

-- Changes from Suck-4.2.5 -> Suck-4.3.0

* As of this version, my e-mail address has changed from
from bobyetman@worldnet.att.net to bobyetman@home.com

* 	changes to allow it to use SSL.

* rpost.c - added -i option, to ignore the
	readonly opening response and try
	to post anyway, since inn-2.3, when
	using authinfo, still sends 201 code
	meaning readonly.

	- do_authenticate() - added print
	statements to show start of, and
	successful completion of authentication.

* suck.c - get_one_article_kill() - add test
	for header_only mode when in stdout mode 
	to not download the body.

	- get_one_article(), get_articles() - added
	 -bP option, to post articles during download.

Sep 17, 2001 - Suck-4.3.0 released

-- Changes from Suck-4.2.4 -> Suck-4.2.5

* batch.c - post_one_msg() - fixed bug in
	testing for a long input line that
	caused us to reject any line that is
	exactly 13 char long.

* chhistory_db.c - various changes to allow it
	to compile with Inn-2.3 headers.

* rpost.c - do_article() - added test for second
	duplicate article string to handle DNews.

	- filter_batch_article() - added option for
	showing the file name as we upload it.

Dec 02, 2000 - Suck-4.2.5 released

-- Changes from Suck-4.2.3 -> Suck-4.2.4

*	various changes to allow compile under OS/2
	(look for ifdef EMX).  You'll need to edit the
	Makefile to get it to compile.

*both.c - true_str(), null_str() - moved from suck.c so
	rpost can use.

* killprg.c - Changes to include so will compile using 
	Perl5.6.  If you're using perl5.004 or older,
	you'll need to define OLD_PERL in the Makefile.

* rpost.c - Changes so will compile with perl5.6.0.
  
	do_article() - Added actual error message from
	remote server to Malfunction error message, to
	make it easier to figure out problem. 

* suck.c - do_one_group() - added code for resetcounter 
	option.  If remote end resets its article numbers
	the normal suck response is to ignore the group
	and reset the lastread counter to match the current
	high counter.  This option tells suck to reset
	its lastread counter to the low counter, effectively
	causing suck to get all articles for the group, and
	use the historydb to check for dupes.
	Added low_read option. This option is used in concert
	with the maxread option of the sucknewsrc.  Normally
	when there is a maxread in the sucknewsrc, suck will
	download the newest articles.  This option tells suck
	to download instead the oldest articles.
	
	do_articles() - added code for show_group option. 
	This option will add the name of the current group
	to the BPS display as you download the articles.
        
May 21, 2000 - Suck-4.2.4 released

-- Changes from Suck-4.2.2 -> Suck-4.2.3

* both.c - signal_block() - Changes to get rid of invalid 
	#ifdef POSIX_SIGNAL causing this routine never to
	get called.

* killfile.c. parse_killfile() - fixed bug in getting perl
        filter filename.  If you didn't specify debug, the
        nl was getting left on the name, causing errors.

* rpost.c - * - Major internal re-write to handle future
	expansion of filters/batch mode.  Changed many
	function's arg passing, so they only pass around
	myargs.  Re-wrote batch mode to make adding new
	filters easier.  Added rnews mode, for reading
	rnews files.  See man page for more details.

* suck.c - main(), sighandler() - changes from signal() to
	sigaction() so I use all POSIX signal handling stuff.

        - do_nodownload() - added, this routine allows you
        to specifiy Message-IDs to never download.  See the
        SUCKNODOWNLOAD section in the man page.

23 Jan 2000 - Suck-4.2.3 released

-- Changes from Suck-4.2.1 -> Suck-4.2.2

* lmove.c match_group(), move_msgs(), scan_args() - Re-wrote
	for new options -h and -s to create hard or symbolic
	links to files crossposted to multiple groups.  Re-wrote
 	code to handle malformed header with spaces either before
	or after the newsgroup names.

	load_active() - added test for duplicate group line.


testhost.c
#include <config.h>

#include <stdio.h>
#include <stdlib.h>
#ifdef HAVE_UNISTD_H
#include <unistd.h>
#endif
#ifdef HAVE_DIRENT_H
# include <dirent.h>
#else
# define dirent direct
# ifdef HAVE_SYS_NDIR_H
#  include <sys/ndir.h>
# endif
# ifdef HAVE_SYS_DIR_H
#  include <sys/dir.h>
# endif
#endif
#include <string.h>
#include <sys/types.h>
#include <netdb.h>
#include <sys/stat.h>
#include <errno.h>
#include <ctype.h>

#ifdef DMALLOC
#include <dmalloc.h>
#endif

#include "both.h"
#include "suck_config.h"
#include "phrases.h"

/* set up defaults */
char **test_phrases=default_test_phrases;
char **both_phrases=default_both_phrases;

#define MAXCMDLEN 128	/* length of longest command to send + its args */
struct {
	const char *command;
	int response;
} cmds[]  = { { "help", 100 }, 
	      { "list", 215 },
	      { "list newsgroups", 215 },
	      { "newgroups", 231},
	      { "list overview.fmt", 215},	      
};

enum { CMD_HELP, CMD_LIST, CMD_DESC, CMD_NEWGRP, CMD_OVERVIEW };
enum { RETVAL_OK = 0, RETVAL_ERROR = -1, RETVAL_BAD_DATES = -2, RETVAL_NOAUTH = -3};

/*-functions----*/
int do_a_command(int, int, FILE *, char *, char *, char *, int, void *);
int check_date_format(char *, char *, char *);
void load_phrases(const char *);
void free_phrases(void);

/*------------------------------------------------------------------*/
int main(int argc, char *argv[]) {

	int sockfd, response, loop, cmd, quiet, mode_reader, do_ssl, retval = RETVAL_OK;
	struct stat sbuf;
	unsigned short int portnr;
	FILE *fptr = stdout;		/* used to print output to */
	void *ssl_struct;

	char *host, *buf, *ptr, *cmdargs, *userid, *passwd;
	char ndate[14];	/* 6 for yymmdd 6 for hhmmss and 1 for space and 1 for null */

	const char *phrases;

	/* set up defaults */
	host = getenv("NNTPSERVER");
	cmd = CMD_HELP;
	loop = 1;
	cmdargs = NULL;
	portnr = DEFAULT_NNRP_PORT;
	passwd = NULL;
	userid = NULL;
	phrases = NULL;
	mode_reader = FALSE;
	do_ssl = FALSE;
	quiet = FALSE;		/* do we show status or not */

	/* have to do this next so if set on cmd line, overrides this */
#ifdef N_PHRASES		/* in case someone nukes def */
	if(stat(N_PHRASES, &sbuf) == 0 && S_ISREG(sbuf.st_mode)) {
		/* we have a regular phrases file make it the default */
		phrases = N_PHRASES;
	}
	
#endif

	if(argc > 1 && argv[loop][0] != '-') {
		host = argv[loop++];
	}
	for(;loop<argc;loop++) {
		if(argv[loop][0] == '-') {
			switch(argv[loop][1]) {
			  case 'a': 
				cmd = CMD_LIST;
				break;
			  case 'd':
				  cmd = CMD_DESC;
				  break;
			  case 'o':
				  cmd = CMD_OVERVIEW;
				  break;
			  case 'q':
				  quiet = TRUE;
				  break;
			  case 'M':
				  mode_reader = TRUE;
				  break;
			  case 'n':
				cmd = CMD_NEWGRP;
				/* next two args must be date & time YYMMDD HHMMSS format */
				if(loop + 2 > argc) {
					error_log(ERRLOG_REPORT, test_phrases[0], NULL);
					retval = RETVAL_ERROR;
				}
				else {
					if((retval = check_date_format(ndate, argv[loop+1], argv[loop+2])) == RETVAL_OK) {
						cmdargs = ndate;
					}
					else {
						error_log(ERRLOG_REPORT, test_phrases[1], argv[loop+1], argv[loop+2], NULL);
					}
					loop += 2;	/* past the args if there, if not this will finish up the arg loop */
				}
				break;
			  case 'e':	/* use default error log path */
				error_log(ERRLOG_SET_FILE, ERROR_LOG, NULL);
				break;
			  case 'E':	/* error log path */
				if(loop+1 == argc) { 
					error_log(ERRLOG_REPORT, test_phrases[2], NULL);
					retval = RETVAL_ERROR;
				}
				else {
					error_log(ERRLOG_SET_FILE, argv[++loop], NULL);
				}
				break;
			  case 's':	/* use default status log path */
				if((fptr = fopen(STATUS_LOG, "a")) == NULL) {
					MyPerror(test_phrases[3]);
					retval = RETVAL_ERROR;
				}
				break;
			  case 'S':	/* status log path */
				if(loop+1 == argc) { 
					error_log(ERRLOG_REPORT, test_phrases[12], NULL);
					retval = RETVAL_ERROR;
				}
				else if((fptr = fopen(argv[++loop], "a")) == NULL) {
					MyPerror(test_phrases[3]);
					retval = RETVAL_ERROR;
				}
				break;
			  case 'N':	/* override port nr */
				if(loop+1 == argc) {
					error_log(ERRLOG_REPORT, test_phrases[4], NULL);
					retval = RETVAL_ERROR;
				}
				else {
					portnr = atoi(argv[++loop]);
				}
				break;
			  case 'U': 	/* next arg is userid for authorization */
				if(loop+1 == argc) {
					error_log(ERRLOG_REPORT, test_phrases[5], NULL);
					retval = RETVAL_ERROR;
				}
				else { 
					userid = argv[++loop];
				}
				break;
			  case 'P': 	/* next arg is password for authorization */
				if(loop+1 == argc) {
					error_log(ERRLOG_REPORT, test_phrases[6], NULL);
					retval = RETVAL_ERROR;
				}
				else { 
					passwd = argv[++loop];
				}
				break;
			  case 'Q':   /* get from env */
				userid = getenv("NNTP_USER");
				passwd = getenv("NNTP_PASS");
				break;
			  case 'l': 	/* next arg is phrases file */
				if(loop+1 == argc) {
					error_log(ERRLOG_REPORT, test_phrases[11], NULL);
					retval = RETVAL_ERROR;
				}
				else { 
					phrases = argv[++loop];
				}
				break;
#ifdef TIMEOUT
			case 'T':  /* timeout */
				if(loop+1 == argc) {
					error_log(ERRLOG_REPORT, test_phrases[13], NULL);
					retval = RETVAL_ERROR;
				}
				else {
					TimeOut = atoi(argv[++loop]);
				}
				break;
#endif
#ifdef HAVE_LIBSSL
			case 'z':
				do_ssl = TRUE;
				portnr = DEFAULT_SSL_PORT;
				break;
#endif
			  default:
				retval = RETVAL_ERROR;
				error_log(ERRLOG_REPORT, test_phrases[7], argv[loop], NULL);
			}
		}
		else {
			retval = RETVAL_ERROR;
			error_log(ERRLOG_REPORT, test_phrases[7], argv[loop], NULL);
		}
	}

	if(retval == RETVAL_OK) {
		load_phrases(phrases);	/* this is here so everything displays okay */

		sockfd = connect_to_nntphost( host, NULL, 0, (quiet == FALSE)?  fptr : NULL, portnr, do_ssl, &ssl_struct);
		if(sockfd < 0 ) {
			retval = RETVAL_ERROR;
		}
		/* get the announcement line */
		else if(sgetline(sockfd, &buf, do_ssl, ssl_struct) < 0) {
			retval = RETVAL_ERROR;
		}
		else {
			ptr = number(buf, &response);
			if(quiet == FALSE) {
				fprintf(fptr,"%s",ptr);	/* print it */
			}
			
			if(response != 200 && response != 201) {
				/* not a valid response */
				retval = RETVAL_ERROR;
			}
			else {
				if(mode_reader == TRUE) {
					sputline(sockfd, "mode reader\r\n", do_ssl, ssl_struct);
					sgetline(sockfd, &buf, do_ssl, ssl_struct);
				}
				do_a_command(sockfd, cmd, fptr, cmdargs, userid, passwd, do_ssl, ssl_struct);
			}
		}
		if(sockfd >= 0) {
			disconnect_from_nntphost(sockfd, do_ssl, &ssl_struct);
		}
		
		if(retval == RETVAL_ERROR) {
			error_log(ERRLOG_REPORT, test_phrases[8], host, NULL);
		}
	}
	if(fptr != stdout && fptr != NULL) {
		fclose(fptr);
	}
	free_phrases();	/* do this last so everything is displayed correctly */
	exit(retval);
}
/*-------------------------------------------------------------------------------*/
int do_a_command(int sockfd, int cmd, FILE *fptr, char *cmdargs, char *userid, char *passwd, int do_ssl, void *ssl_struct ) {
	char *ptr, *buf, cmdstr[MAXCMDLEN], buf2[MAXCMDLEN];
	int response;
	int len, done, retval = RETVAL_OK;

	/* build command to send */
	strcpy(cmdstr, cmds[cmd].command);
	if(cmdargs != NULL) {
		strcat(cmdstr, " ");
		strcat(cmdstr, cmdargs);
	}
	strcat(cmdstr, "\r\n");
	
	sputline(sockfd, cmdstr, do_ssl, ssl_struct);		/* which command do I run? */
	sgetline(sockfd, &buf, do_ssl, ssl_struct);			/* get response */

	ptr = number(buf, &response);
	if(response == 480 ) { /* we must do authorization */
		sprintf(buf2, "AUTHINFO USER %s\r\n", userid);
		sputline(sockfd, buf2, do_ssl, ssl_struct);
		len = sgetline(sockfd, &buf, do_ssl, ssl_struct);
		if( len < 0) {	
			retval = RETVAL_ERROR;		  	
		}					
		else {
			ptr  = number(buf, &response);
			if(response != 381) {
				error_log(ERRLOG_REPORT, test_phrases[9], buf, NULL);
				retval = RETVAL_NOAUTH;
			}
			else {
				sprintf(buf2, "AUTHINFO PASS %s\r\n", passwd);
				sputline(sockfd, buf2, do_ssl, ssl_struct);
				len = sgetline(sockfd, &buf, do_ssl, ssl_struct);
				if(len < 0) {	
					retval = RETVAL_ERROR;		  	
				}					
				else {
					number(buf, &response);
					switch(response) {
					  case 281: /* bingo resend original command*/
						sputline(sockfd, cmdstr, do_ssl, ssl_struct);
						len = sgetline(sockfd, &buf, do_ssl, ssl_struct);
						if( len < 0) {	
							retval = RETVAL_ERROR;		  	
						}
						else {
							ptr = number(buf,&response);
						}
						break;
					  case 502: /* permission denied */
						retval = RETVAL_NOAUTH;
						error_log(ERRLOG_REPORT, test_phrases[10], NULL);
						break;
					  default: /* wacko error */
						error_log(ERRLOG_REPORT, test_phrases[9], ptr, NULL);
						retval = RETVAL_NOAUTH;
						break;
					}
				}
			}
		}
	}
	if(cmds[cmd].response != response) {
		error_log(ERRLOG_REPORT, "%v1%\n", buf, NULL);
		retval = RETVAL_ERROR;
	}
	else {
		/* okay here we go */
		fprintf(fptr, "%s", ptr); 
		/* commands are ended by a . on a line by itself */

		len = done = 0;
		while( len >=0 && done == 0) {
		
			len = sgetline(sockfd, &buf, do_ssl, ssl_struct);

			if(len == 2 && strcmp(buf, ".\n") == 0) {
				done = 1;
			}
			else if(len > 0) {
				fprintf(fptr,"%s",buf);
			}
		}
		if(len < 0) {
			retval = RETVAL_ERROR;
		}
	}
	return retval;
}
/*---------------------------------------------------------------*/
int check_date_format(char *dates, char *indate, char *intime) {

	/* if indate & intime are not valid format, return error */
	/* when done, dates will be yymmdd hhmmss */
	
	int i, retval = RETVAL_OK;	
	
	/* now test my incoming args */
	if(indate == NULL || intime == NULL) {
		retval = RETVAL_ERROR;
	}
	else if(strlen(indate) != 6 || strlen(intime) != 6) {
		retval = RETVAL_ERROR;
	}
	else {
		for(i=0;i<6;i++) {
			if(!isdigit(indate[i]) || !isdigit(intime[i])) {
					retval = RETVAL_ERROR;
			}
		}
	}

	if(retval == RETVAL_OK) {
		sprintf(dates, "%s %s", indate, intime);
	}
	return retval;
}
/*--------------------------------------------------------------------------------*/
/* THE strings in this routine is the only one not in the arrays, since           */
/* we are in the middle of reading the arrays, and they may or may not be valid.  */
/*--------------------------------------------------------------------------------*/
void load_phrases(const char *phrases) {

	int error=TRUE;
	FILE *fpi;
	char buf[MAXLINLEN];
	
	if(phrases != NULL) {
		
		if((fpi = fopen(phrases, "r")) == NULL) {
			MyPerror(phrases);
		}
		else {
			fgets(buf, MAXLINLEN, fpi);
			if(strncmp( buf, SUCK_VERSION, strlen(SUCK_VERSION)) != 0) {
				error_log(ERRLOG_REPORT, "Invalid Phrase File, wrong version\n", NULL);
			}
			else if((both_phrases = read_array(fpi, NR_BOTH_PHRASES, TRUE)) != NULL) {
				read_array(fpi, NR_RPOST_PHRASES, FALSE);	/* skip these */
				if((test_phrases = read_array(fpi, NR_TEST_PHRASES, TRUE)) != NULL) {
					error = FALSE;
				}
			}
		}
		fclose(fpi);
		if(error == TRUE) {
			/* reset back to default */
			error_log(ERRLOG_REPORT, "Using default Language phrases\n", NULL);
			test_phrases = default_test_phrases;
			both_phrases = default_both_phrases;
		}
	}		
}
/*--------------------------------------------------------------------------------*/
void free_phrases(void) {
		/* free up the memory alloced in load_phrases() */
		if(test_phrases != default_test_phrases) {
			free_array(NR_TEST_PHRASES, test_phrases);
		}
		if(both_phrases != default_both_phrases) {
			free_array(NR_BOTH_PHRASES, both_phrases);
		}
		
}
